File: Searching and Sorting/Merge sort/C#/merge_sort.cs
int[] arr = new int[10]
{
    1, 5, 4, 11, 20, 8, 2, 98, 90, 16
};
 
MergeSort(arr, 0, arr.Length - 1);
Console.WriteLine("Sorted Values:");
for (int i = 0; i < arr.Length; i++)
    Console.WriteLine(arr[i]);


private void Merge(int[] input, int left, int middle, int right)
{
    int[] leftArray = new int[middle - left + 1];
    int[] rightArray = new int[right - middle];
 
    Array.Copy(input, left, leftArray, 0, middle - left + 1);
    Array.Copy(input, middle + 1, rightArray, 0, right - middle);
 
    int i = 0;
    int j = 0;
    for (int k = left; k < right + 1; k++)
    {
        if (i == leftArray.Length)
        {
            input[k] = rightArray[j];
            j++;
        }
        else if (j == rightArray.Length)
        {
            input[k] = leftArray[i];
            i++;
        }
        else if (leftArray[i] <= rightArray[j])
        {
            input[k] = leftArray[i];
            i++;
        }
        else
        {
            input[k] = rightArray[j];
            j++;
        }
    }
}
 
private void MergeSort(int[] input, int left, int right)
{
    if (left < right)
    {
        int middle = (left + right) / 2;
 
        MergeSort(input, left, middle);
        MergeSort(input, middle + 1, right);
 
        Merge(input, left, middle, right);
    }
}

File: Searching and Sorting/Merge sort/C#/mergesort.cpp
#include<bits/stdc++.h>
using namespace std;

void merge(int arr[],int beg,int mid,int end)
{
    int a[10];
    int k=0,i=beg,j=mid+1;

    while(i<=mid&&j<=end)
    {
        if(arr[i]<arr[j])
        {
            a[k]=arr[i];
            k++;
            i++;
        }
        else
        {
            a[k]=arr[j];
            k++;
            j++;
        }
    }

    if(i>mid)
    {
        while(j<=end)
        {
            a[k]=arr[j];
            k++;
            j++;
        }
    }
    else
    {
        while(i<=mid)
        {
            a[k]=arr[i];
            k++;
            i++;
        }
    }
    int l=0;
    for(int i=beg;i<=end;i++)
    {
      arr[i]=a[l];
      l++;
    }

}


void mergesort(int arr[],int beg,int end)
{
    if(beg<end)
    {
        int mid=(beg+end)/2;
        mergesort(arr,beg,mid);
        mergesort(arr,mid+1,end);
        merge(arr,beg,mid,end);
    }

}

int main()
{
    int arr[10]={5,3,7,9,11,8,2,4,10,6};

    mergesort(arr,0,9);
    for(int i=0;i<10;i++)
        cout<<arr[i]<<endl;
    return 0;

}
